File: samples/fork-bomb-c/bot.c
/* Tier-3 resource-abuse sample: forks children without bound on the first
 * move. Only meaningful (safely containable) under the hardened engine's
 * PidsLimit; do not run this outside it. */
#include <unistd.h>

#include "bot.h"

int choose_move(const int board[COLS][ROWS], int you, const MoveInfo *info) {
    (void)you;
    (void)info;
    for (;;) {
        if (fork() < 0) break;
    }
    int moves[COLS];
    int n = legal_moves(board, moves);
    return n > 0 ? moves[0] : 0;
}

int legal_moves(const int board[COLS][ROWS], int out_cols[COLS]) {
    int n = 0;
    for (int col = 0; col < COLS; col++) {
        if (board[col][ROWS - 1] == 0) {
            out_cols[n++] = col;
        }
    }
    return n;
}

File: samples/cpu-spin-python/bot.py
"""Tier-3 resource-abuse sample: a background thread spins the CPU while the
bot still answers /move legally on the request thread. Only meaningful under
the hardened engine's 1-CPU cap; do not run this outside it."""

import threading


def _spin():
    while True:
        pass


threading.Thread(target=_spin, daemon=True).start()


def choose_move(board, you, info=None):
    return legal_moves(board)[0]


def legal_moves(board):
    return [col for col, column in enumerate(board) if column[-1] == 0]
